File: shared/src/main/java/xyz.lychee.dynamicdns/shared/MessageReceiver.java
package xyz.lychee.dynamicdns.shared;

import dev.dejvokep.boostedyaml.YamlDocument;
import lombok.Getter;

import javax.crypto.Cipher;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.spec.SecretKeySpec;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.logging.Level;
import java.util.logging.Logger;

@Getter
public abstract class MessageReceiver {
    private final Map<String, InetSocketAddress> servers = new ConcurrentHashMap<>();
    private final YamlDocument config;
    private final Logger logger;
    private Cipher cipher;

    public MessageReceiver(Logger logger, YamlDocument config) {
        this.config = config;
        this.logger = logger;
    }

    public static Cipher createCipher(String token, int mode) throws NoSuchPaddingException, NoSuchAlgorithmException, InvalidKeyException {
        byte[] keyBytes = new byte[16];
        byte[] tokenBytes = token.getBytes();
        System.arraycopy(tokenBytes, 0, keyBytes, 0, Math.min(tokenBytes.length, keyBytes.length));

        Cipher cipher = Cipher.getInstance("AES");
        cipher.init(mode, new SecretKeySpec(keyBytes, "AES"));
        return cipher;
    }

    public void loadServers(String token) {
        try {
            this.cipher = MessageReceiver.createCipher(token, Cipher.DECRYPT_MODE);
        } catch (NoSuchPaddingException | NoSuchAlgorithmException | InvalidKeyException e) {
            this.logger.log(Level.SEVERE, "Failed to load config", e);
        }

        List<String> servers = this.config.getStringList("servers");
        if (servers != null && !servers.isEmpty()) {
            for (String server : servers) {
                String[] split = server.split("-");
                this.servers.put(split[0], AddressUtil.parseAddress(split[1]));
            }
        }
    }

    private void saveServers() throws IOException {
        List<String> list = new ArrayList<>();
        for (Map.Entry<String, InetSocketAddress> entry : this.servers.entrySet()) {
            String name = entry.getKey();
            InetSocketAddress address = entry.getValue();
            list.add(name + '-' + address.getHostString() + ':' + address.getPort());
        }

        this.config.set("servers", list);
        this.config.save();
    }

    public abstract void updateServers();

    public void start(InetSocketAddress address) {
        Thread t = new Thread(() -> {
            try (ServerSocket serverSocket = new ServerSocket(address.getPort())) {
                this.logger.info("Started listening " + address + " for ip changes!");

                while (true) {
                    try (Socket clientSocket = serverSocket.accept()) {
                        clientSocket.setSoTimeout(5000);

                        byte[] decrypted = this.cipher.doFinal(this.readAllBytes(clientSocket.getInputStream()));
                        String message = new String(decrypted, StandardCharsets.UTF_8);

                        String[] data = message.split(":");

                        if (data.length >= 2) {
                            String serverName = data[0];
                            int serverPort = Integer.parseInt(data[1]);
                            InetSocketAddress serverAddress = new InetSocketAddress(clientSocket.getInetAddress(), serverPort);

                            this.servers.put(serverName, serverAddress);

                            this.updateServers();
                            this.saveServers();
                        } else {
                            this.logger.warning("Invalid message format: " + message);
                        }
                    } catch (Exception ex) {
                        this.logger.warning("An error occurred while reading the message: " + ex.getMessage());
                    }
                }
            } catch (IOException ex) {
                this.logger.log(Level.SEVERE, "Failed to start UDP server", ex);
            }
        });

        t.setName("DynamicDNS-Server");
        t.start();
    }

    private byte[] readAllBytes(InputStream inputStream) throws IOException {
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        byte[] data = new byte[8192];
        int bytesRead;

        while ((bytesRead = inputStream.read(data, 0, data.length)) != -1) {
            buffer.write(data, 0, bytesRead);
        }

        return buffer.toByteArray();
    }
}

File: shared/src/main/kotlin/xyz/lychee/dynamicdns/shared/MessageReceiver.kt
package xyz.lychee.dynamicdns.shared

import dev.dejvokep.boostedyaml.YamlDocument
import dev.dejvokep.boostedyaml.block.implementation.Section
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.io.InputStream
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.nio.charset.StandardCharsets
import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.logging.Level
import java.util.logging.Logger
import javax.crypto.Cipher
import javax.crypto.NoSuchPaddingException
import javax.crypto.spec.SecretKeySpec

abstract class MessageReceiver(
    val logger: Logger,
    val config: YamlDocument
) {
    val servers: MutableMap<String, InetSocketAddress> = ConcurrentHashMap()
    val blacklist: MutableSet<InetAddress> = HashSet()
    val executor: ExecutorService = Executors.newCachedThreadPool { r ->
        Thread(r).apply { name = "DynamicDNS-Worker" }
    }
    var cipher: Cipher? = null

    companion object {
        @JvmStatic
        @Throws(NoSuchPaddingException::class, NoSuchAlgorithmException::class, InvalidKeyException::class)
        fun createCipher(token: String, mode: Int): Cipher {
            val keyBytes = ByteArray(16)
            val tokenBytes = token.toByteArray()
            System.arraycopy(tokenBytes, 0, keyBytes, 0, minOf(tokenBytes.size, keyBytes.size))

            val cipher = Cipher.getInstance("AES")
            cipher.init(mode, SecretKeySpec(keyBytes, "AES"))
            return cipher
        }
    }

    fun loadServers(token: String) {
        try {
            this.cipher = createCipher(token, Cipher.DECRYPT_MODE)
        } catch (e: Exception) {
            this.logger.log(Level.SEVERE, "Failed to load config", e)
        }

        if (!this.config.contains("servers")) return

        if (this.config.isSection("servers")) {
            val section: Section? = this.config.getSection("servers")
            if (section != null) {
                for (key in section.keys) {
                    val serverName = key.toString()
                    this.servers[serverName] = AddressUtil.parseAddress(section.getString(serverName))
                }
            }
        } else if (this.config.isList("servers")) {
            val serversList = this.config.getStringList("servers")
            if (!serversList.isNullOrEmpty()) {
                for (server in serversList) {
                    val split = server.split("-")
                    this.servers[split[0]] = AddressUtil.parseAddress(split[1])
                }
            }
        }
    }

    @Throws(IOException::class)
    protected open fun saveServers() {
        this.servers.forEach { (name, address) ->
            this.config.set("servers.$name", "${address.hostString}:${address.port}")
        }
        this.config.save()
    }

    abstract fun updateServers()

    fun start(address: InetSocketAddress) {
        val mainThread = Thread {
            try {
                ServerSocket(address.port).use { serverSocket ->
                    this.logger.info("Started listening $address for ip changes!")

                    while (!Thread.currentThread().isInterrupted) {
                        try {
                            val clientSocket = serverSocket.accept()
                            val inetAddress = clientSocket.inetAddress

                            if (inetAddress == null || this.blacklist.contains(inetAddress)) {
                                this.silentClose(clientSocket)
                                continue
                            }

                            this.executor.submit { this.handleClient(clientSocket, inetAddress) }
                        } catch (ex: IOException) {
                            if (serverSocket.isClosed) break
                            this.logger.log(Level.WARNING, "Error accepting connection", ex)
                        }
                    }
                }
            } catch (ex: IOException) {
                this.logger.log(Level.SEVERE, "Failed to start TCP server", ex)
            } finally {
                executor.shutdown()
            }
        }

        mainThread.name = "DynamicDNS-Server"
        mainThread.start()
    }

    private fun handleClient(clientSocket: Socket, inetAddress: InetAddress) {
        try {
            clientSocket.soTimeout = 3000

            val rawBytes = this.readAllBytesBounded(clientSocket.getInputStream())
            if (rawBytes.isEmpty()) {
                this.silentClose(clientSocket)
                return
            }

            val decrypted = this.cipher!!.doFinal(rawBytes)
            val message = String(decrypted, StandardCharsets.UTF_8)

            val data = message.split(":", limit = 3)

            if (data.size >= 2) {
                val serverName = data[0]
                val serverPort = data[1].trim().toInt()
                val serverAddress = InetSocketAddress(inetAddress, serverPort)

                this.servers[serverName] = serverAddress

                this.updateServers()
                this.saveServers()
            } else {
                this.logger.warning("Invalid message format from ${inetAddress.hostAddress}")
                this.blacklist.add(inetAddress)
                this.silentClose(clientSocket)
            }
        } catch (ex: Exception) {
            if (this.logger.isLoggable(Level.FINE)) {
                this.logger.log(Level.FINE, "Bot/Scanner connection issue from ${inetAddress.hostAddress}", ex)
            }

            this.blacklist.add(inetAddress)
            this.silentClose(clientSocket)
        }
    }

    @Throws(IOException::class)
    private fun readAllBytesBounded(input: InputStream): ByteArray {
        val buffer = ByteArrayOutputStream()
        val data = ByteArray(512)
        var bytesRead: Int
        var totalRead = 0

        while (input.read(data, 0, data.size).also { bytesRead = it } != -1) {
            buffer.write(data, 0, bytesRead)
            totalRead += bytesRead
            if (totalRead > 2048) {
                throw IOException("Packet size limit exceeded (DoS Protection)")
            }
        }
        return buffer.toByteArray()
    }

    fun silentClose(clientSocket: Socket?) {
        try {
            if (clientSocket != null && !clientSocket.isClosed) {
                clientSocket.setSoLinger(true, 0)
                clientSocket.close()
            }
        } catch (ignored: IOException) {
        }
    }
}
